src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -537,14 +539,38 @@ protected override void CreateIndicesForNonTypeMembers(ITypeDefinition typeDef)
 
             foreach (var methodDef in typeDef.GetMethods(this.Context))
             {
-                if (this.AddDefIfNecessary(_methodDefs, methodDef))
+                this.AddDefIfNecessary(_methodDefs, methodDef);
+                var methodChange = _changes.GetChange(methodDef);
+
+                if (methodChange == SymbolChange.Added)
                 {
+                    _firstParamRowMap.Add(GetMethodDefinitionHandle(methodDef), _parameterDefs.NextRowId);
                     foreach (var paramDef in this.GetParametersToEmit(methodDef))
                     {
                         _parameterDefs.Add(paramDef);
-                        _parameterDefList.Add(KeyValuePairUtil.Create(methodDef, paramDef));
+                        _parameterDefList.Add(paramDef, methodDef);
+                    }
+                }
+                else if (methodChange == SymbolChange.Updated)
+                {
+                    // If we're re-emitting parameters for an existing method we need to find their original row numbers
+                    // and reuse them so the EnCLog, EnCMap and CustomAttributes tables refer to the right rows
+
+                    // Unfortunately we have to check the original metadata and deltas separately as nothing tracks the aggregate data
+                    // in a way that we can use
+                    var handle = GetMethodDefinitionHandle(methodDef);
+                    if (_previousGeneration.OriginalMetadata.MetadataReader.GetTableRowCount(TableIndex.MethodDef) >= MetadataTokens.GetRowNumber(handle))
+                    {
+                        EmitParametersFromOriginalMetadata(methodDef, handle);
+                    }
+                    else
+                    {
+                        EmitParametersFromDelta(methodDef, handle);
                     }
+                }
 
+                if (methodChange == SymbolChange.Added)
```

Comment:
I'm assuming the behavior of type parameters is unchanged here, i.e. adding an attribute to a type parameter during an EnC session is not supported.

Reply:
Yes, editing generic types is not supported.